hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioOutboundPipeline.java

```diff
@@ -140,14 +156,70 @@ public void write(OutboundFrame frame) {
         } else {
             writeQueue.offer(frame);
         }
-        schedule();
+
+        // take care of the scheduling.
+        for (; ; ) {
+            State state = scheduled.get();
+            if (state == State.UNSCHEDULED) { // pipeline isn't scheduled, so we need to schedule it.
+                if (!scheduled.compareAndSet(State.UNSCHEDULED, State.SCHEDULED)) {
+                    continue; // try again
+                }
+
+                // we manage to schedule the pipeline, meaning we are owner.
+                if (writeThroughEnabled && !concurrencyDetection.isDetected()) {
+                    // we are allowed to do a write through, so lets process the request on the calling thread
+                    try {
+                        process();
+                    } catch (Throwable t) {
+                        onError(t);
+                    }
+                } else {
+                    ownerAddTaskAndWakeup(this); // no write through, let the io thread deal with it.
+                }
+                return;
+            } else if (state == State.SCHEDULED || state == State.RESCHEDULE) { // already scheduled, so we are done
+                if (writeThroughEnabled) {
+                    concurrencyDetection.onDetected();
+                }
+                return;
+            } else if (state == State.BLOCKED) { // pipeline is blocked, so we don't need to schedule
+                return;
+            } else {
+                throw new IllegalStateException("Unexpected state:" + state);
+            }
+        }
+    }
+
+    @Override
+    public OutboundPipeline wakeup(String reason) {
+        //System.out.println(getChannel() + " wakeup: '" + reason + "', current " + scheduled.get() + " ");
+
+        for (; ; ) {
+            State prevState = scheduled.get();
+            if (prevState == State.RESCHEDULE) {
+                break;
+            } else {
+                // the pipeline is already scheduled, it is concern of the current owner to check if the pipeline
```

Comment:
`the pipeline is already scheduled` -> is this in fact true? I believe at this point the pipeline might be unscheduled and we are about to reschedule it.

Reply:
You are right. The documentation isn't correct; I guess it is a copy paste leftover.